keps/sig-node/0028-20180906-quotas-for-ephemeral-storage.md

```diff
@@ -0,0 +1,647 @@
+filesystem consumption that can be applied to one or more directories.
+
+## Motivation
+
+The mechanism presently used to monitor storage consumption involves
+use of `du` and `find` to periodically gather information about
+storage and inode consumption of volumes.  This mechanism suffers from
+a number of drawbacks:
+
+* It is slow.  If a volume contains a large number of files, walking
+  the directory can take a significant amount of time.  There has been
+  at least one known report of nodes becoming not ready due to volume
+  metrics: <https://github.com/kubernetes/kubernetes/issues/62917>
+* It is possible to conceal a file from the walker by creating it and
+  removing it while holding an open file descriptor on it.  POSIX
+  behavior is to not remove the file until the last open file
+  descriptor pointing to it is removed.  This has legitimate uses; it
+  ensures that a temporary file is deleted when the processes using it
+  exit, and it minimizes the attack surface by not having a file that
+  can be found by an attacker.  The following pod does this; it will
+  never be caught by the present mechanism:
+
+```yaml
+apiVersion: v1
+kind: Pod
+max:
+metadata:
+  name: "diskhog"
+spec:
+  containers:
+  - name: "perl"
+    resources:
+      limits:
+        ephemeral-storage: "2048Ki"
+    image: "perl"
+    command:
+    - perl
+    - -e
+    - >
+      my $file = "/data/a/a"; open OUT, ">$file" or die "Cannot open $file: $!\n"; unlink "$file" or die "cannot unlink $file: $!\n"; my $a="0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"; foreach my $i (0..200000000) { print OUT $a; }; sleep 999999
+    volumeMounts:
+    - name: a
+      mountPath: /data/a
+  volumes:
+  - name: a
+    emptyDir: {}
+```
+* It is reactive rather than proactive.  It does not prevent a pod
+  from overshooting its limit; at best it catches it after the fact.
+  On a fast storage medium, such as NVMe, a pod may write 50 GB or
+  more of data before the housekeeping performed once per minute
+  catches up to it.  If the primary volume is the root partition, this
+  will completely fill the partition, possibly causing serious
+  problems elsewhere on the system.
+
+In many environments, these issues may not matter, but shared
+multi-tenant environments need these issues addressed.
+
+### Goals
+
+* Primary: improve performance of monitoring by using project quotas
+  in a non-enforcing way to collect information about storage
+  utilization.
+* Primary: detect storage used by pods that is concealed by deleted
+  files being held open.
+* Primary: this will not interfere with the more common user and group
+  quotas.
+* Stretch: enforce limits on per-volume storage consumption by using
+  enforced project quotas.  Each volume would be given an enforced
+  quota of the total ephemeral storage limit of the pod.
+
+### Non-Goals
+
+* Enforcing limits on total pod storage consumption by any means, such
```

Comment:
"Enforcing Node Allocatable" would mean enforcing a limit over the sum of all pod's disk usage.  For memory, for example, this means the `/kubepods` cgroup has a memory limit of allocatable.  The purpose of it is to protect system daemons even in the case where containers do not set limits.  In this case, it would mean that the abuse vectors you point out cannot cause the node to run out of disk, but at worst cause temporary problems for other pods.  However, currently our image accounting design has problems (images are considered node "overhead", which makes setting kube-reserved rather strange), so we probably don't want to enforce Node Allocatable until that is solved as well.

cc @yujuhong, who drove the initial CRI design and improvements.

Reply:
This is strictly referring to local storage capacity isolation aka ephemeral storage, at least at this point.

I will add "enforcing node allocatable" to the non-goals and make clear that the goals apply only to ephemeral storage.

Reply:
node allocatable exists for ephemeral storage, we just don't want to enforce it anytime soon.

Reply:
Adding non-goal.